torch/distributions/bernoulli.py

```diff
@@ -37,20 +45,17 @@ def sample(self, sample_shape=torch.Size()):
         shape = self._extended_shape(sample_shape)
         return torch.bernoulli(self.probs.expand(shape))
 
+    def _binary_cross_entropy(self, value):
+        max_val = (-self.logits).clamp(min=0)
+        return self.logits - self.logits * value + max_val + \
+            ((-max_val).exp() + (-self.logits - max_val).exp()).log()
+
     def log_prob(self, value):
         self._validate_log_prob_arg(value)
-        param_shape = value.size()
-        probs = self.probs.expand(param_shape)
-        # compute the log probabilities for 0 and 1
-        log_pmf = (torch.stack([1 - probs, probs], dim=-1)).log()
-        # evaluate using the values
-        return log_pmf.gather(-1, value.unsqueeze(-1).long()).squeeze(-1)
+        return -self._binary_cross_entropy(value)
```

Comment:
Is it still numerically stable if we use the [non-reducing C implementation](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1235) with `probs` instead of `logits`?
```py
return -torch.nn.functiona.binary_cross_entropy(self.probs, value, reduce=False)
```
(Note that `torch.nn.functional.binary_cross_entropy_with_logits()` does not allow `reduce=False`)

Reply:
I wanted to use `cross_entropy_with_logits` but you are right that `reduce` arg is not available with that. 

If we are operating with logits, this will be more stable. See:

```
>>> b = Bernoulli(logits=torch.FloatTensor([-32, -8]))
>>> b.log_prob(torch.FloatTensor([0, 1]))

-0.0000
-8.0003
[torch.FloatTensor of size 2]

>>> torch._C._nn.binary_cross_entropy(Variable(b.probs), Variable(torch.FloatTensor([0, 1])), reduce=False)
Variable containing:
-9.8743e-13
 8.0003e+00
[torch.FloatTensor of size 2]
```

Reply:
Great, just checking. It might be cleaner to add a `reduce=True` argument to `torch.nn.functional.binary_cross_entropy_with_logits()` now that we're moving this code into PyTorch.

Reply:
Good idea, will do that instead.